### PR TITLE
[JENKINS-69491] Implement support for "@/path/to/spec" for "poll view" in ManualWorkspaceImpl.java too

### PR DESCRIPTION
**Pull request by lleroy**
Also support "@" in the view for polling.
This is a continuation of what was implemented for JENKINS-69491, but the polling view was forgotten.

### Testing done

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Comment by lleroy** [comment]
@p4paul : could you have a look to have this (rather trivial) fix merged? This is a continuation of a previous change that was merged.
I need it on my production server and now have to use a custom-build version of the plugin, but would like to use an official (future) release

**Comment by skumar7322** [comment]
Thanks @lleroy  for creating this pull request. We will plan this in the upcoming release.

**Comment by skumar7322** [comment]
Hi @lleroy, in the description you mentioned that you want to implement the change in the poll view. Polling will use the mapping specified in the View Mapping which is handled. But you made changes in the Change View, which restricts the file version to a specific version. 

https://www.perforce.com/manuals/p4guide/Content/P4Guide/configuration.workspace_view.changeview.html

File: src/main/java/org/jenkinsci/plugins/p4/workspace/ManualWorkspaceImpl.java [hunk 1]
```diff
@@ -172,12 +172,34 @@ else if (client != null && adjustClientName) {
...
 
-	private ArrayList<String> getChangeView(WorkspaceSpec workspaceSpec) {
+	private ArrayList<String> getChangeView(IOptionsServer connection, WorkspaceSpec workspaceSpec) throws Exception
+    {
 		ArrayList<String> changeView = new ArrayList<>();
 
 		String view = workspaceSpec.getChangeView();
 		if (view != null) {
 			String specString = getExpand().format(view, true);
+			boolean fromFile = false;
```
Comment (skumar7322) [inline src/main/java/org/jenkinsci/plugins/p4/workspace/ManualWorkspaceImpl.java:182]: fromFile variable is never used

File: src/main/java/org/jenkinsci/plugins/p4/workspace/ManualWorkspaceImpl.java [hunk 2]
```diff
@@ -172,12 +172,34 @@ else if (client != null && adjustClientName) {
...
+				// extract View from file:	JENKINS-69491.
+				fromFile = true;
+				logger.fine("P4: view from file=" + specString);
+				String specPathFull = specString.substring(1).trim();
+				List<IFileSpec> file = FileSpecBuilder.makeFileSpecList(specPathFull);
+				GetFileContentsOptions printOpts = new GetFileContentsOptions();
+				printOpts.setNoHeaderLine(true);
+				InputStream ins = null;
+				try {
+					connection.getFileContents(file, printOpts);
```
Comment (skumar7322) [inline src/main/java/org/jenkinsci/plugins/p4/workspace/ManualWorkspaceImpl.java:193]: file content is never assigned, so ins will always be null.

File: src/main/java/org/jenkinsci/plugins/p4/workspace/ManualWorkspaceImpl.java [hunk 3]
```diff
@@ -172,12 +172,34 @@ else if (client != null && adjustClientName) {
...
+				} finally {
+					if (ins != null) {
+						ins.close();
+					}
+				}
+				specString = getExpand().format(specString, true);
+			}
+
 			for (String line : specString.split("\\n")) {
 				changeView.add(line);
```
Comment (skumar7322) [inline src/main/java/org/jenkinsci/plugins/p4/workspace/ManualWorkspaceImpl.java:204]: ClientView in the file can have variables which we need to expand. Better expend it before adding to change view

**Comment by skumar7322** [comment]
> the functionality is that if the view spec is something like "@//depot/path/server.cws" then latest version of the file //depot/path/server.cws is taken from perforce, and that is the viewspec.
> 
> the meat of the implementation was done in #130
> 
> This pr is to also do this for the polling view.
> 
> It does not interfere with the //depot/path/...@1000 - note the startsWith("@")

The method "getChangeView(...)", takes the ClientView from the workspace. And then add it to the ClientView in the perforce Client. Then perforce restrict to the version specified in the mapping for the files matching the specified path. If we do the chnages in this method then it will restrict the version of the file. 

Your change will allow p4-plugin to read the "changeView" mapping from a files available in the perforce instead of the textbox.

File: src/main/java/org/jenkinsci/plugins/p4/workspace/ManualWorkspaceImpl.java (hunk 2 above)
Comment (lleroy) [reply src/main/java/org/jenkinsci/plugins/p4/workspace/ManualWorkspaceImpl.java:193]: yes, line 194 should have been the following of course - guess I forgot to push that before creating the PR
ins = connection.getFileContents(file, printOpts);

File: src/main/java/org/jenkinsci/plugins/p4/workspace/ManualWorkspaceImpl.java (hunk 3 above)
Comment (lleroy) [reply src/main/java/org/jenkinsci/plugins/p4/workspace/ManualWorkspaceImpl.java:204]: no, the variables would be inside the content of the file, not inside the name of the file

**Comment by lleroy** [comment]
> Your change will allow p4-plugin to read the "changeView" mapping from a files available in the perforce instead of the textbox.
Is this what you need?

yes, indeed, as in https://github.com/jenkinsci/p4-plugin/pull/130 (Paul added some documentation somewhere too for that "feature")

**Comment by lleroy** [comment]
I committed the change for the "ins = " but I do not know github enough to update the pull request to reflect that

File: src/main/java/org/jenkinsci/plugins/p4/workspace/ManualWorkspaceImpl.java (hunk 1 above)
Comment (skumar7322) [reply src/main/java/org/jenkinsci/plugins/p4/workspace/ManualWorkspaceImpl.java:182]: pls resolve this comment

Comment (lleroy) [reply src/main/java/org/jenkinsci/plugins/p4/workspace/ManualWorkspaceImpl.java]: yes, indeed. this was copied from JENKINS-69491 where it was used.
fixed in new commit

**Review by lleroy — COMMENTED** [review]
fixed the remarks by 2 new submits on branch

**Pull request merged**